streaming/storage/deque.go

```diff
@@ -0,0 +1,392 @@
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "something went wrong during peek")
+	}
+
+	key := getIndexKey(index)
+
+	err = dq.tx.Delete(key)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete element from queue in pop")
+	}
+
+	err = modifier()
+	if err != nil {
+		return errors.Wrap(err, "failed to actualize queue attributes in pop")
+	}
+
+	return nil
+}
+
+//Returns the first element of the queue without removing it
+//Returns ErrNotFound if the queue is empty
+func (dq *Deque) PeekFront(value proto.Message) error {
+	if !dq.initializedFront {
+		err := dq.initializeFront()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PeekFront")
+		}
+	}
+
+	return dq.peekIndex(value, dq.firstFreeFrontSpot+1)
+}
+
+//Returns the last element of the queue without removing it
+//Returns ErrNotFound if the queue is empty
+func (dq *Deque) PeekBack(value proto.Message) error {
+	if !dq.initializedBack {
+		err := dq.initializeBack()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PeekBack")
+		}
+	}
+
+	return dq.peekIndex(value, dq.firstFreeBackSpot-1)
+}
+
+//we don't check if the queue has been initialized here, since we only
+//call peekIndex from PeekFront/PeekBack, which both initialize the queue
+func (dq *Deque) peekIndex(value proto.Message, index int) error {
+	key := getIndexKey(index)
+
+	data, err := dq.tx.Get(key)
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "an error occurred during Get")
+	}
+
+	err = proto.Unmarshal(data, value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't unmarshal the element in peek")
+	}
+
+	return nil
+}
+
+//Clears all contents of the queue including the metadata
+func (dq *Deque) Clear() error {
+	for dq.firstFreeBackSpot-dq.firstFreeFrontSpot > 1 {
+		key := getIndexKey(dq.firstFreeFrontSpot + 1)
+		err := dq.tx.Delete(key)
+
+		if err != nil {
+			return errors.Wrap(err, "couldn't clear the list")
+		}
+
+		//we call increaseFirstFree instead of simply dq.firstFree += 1 because
+		//a machine can die in the middle of a clear
+		err = dq.increaseFirstFreeFrontSpot()
+		if err != nil {
+			return errors.Wrap(err, "couldn't increase first element index")
+		}
+	}
+
+	err := dq.tx.Delete(dequeFirstFreeFrontSpot)
+	if err != nil {
+		return errors.Wrap(err, "couldn't clear first element index metadata")
+	}
+
+	err = dq.tx.Delete(dequeFirstFreeBackSpot)
+	if err != nil {
+		return errors.Wrap(err, "couldn't clear last element index metadata")
+	}
+
+	/* Generally if someone uses Clear() on the a queue (as in a namespace), then they shouldn't
+	use the same one to add new values. In case they do we can set these to defaults.
+	*/
+	dq.firstFreeFrontSpot = defaultFirstFreeFrontSpot
+	dq.firstFreeBackSpot = defaultFirstFreeBackSpot
+	dq.initializedFront = false
+
+	return nil
+}
+
+func (dq *Deque) GetIterator() *DequeIterator {
```

Comment:
For a Deque it would make sense to provide a reverse iterator too.

Reply:
I added the opts, so you can do it with WithReverse() now